File: ios/Helix/Services/InstanceService.swift
/**
 * Instance Service - Helix iOS
 * Manages instance creation and fetching from Supabase
 */

import Foundation

struct Instance: Codable, Identifiable {
    let id: String
    let userId: String
    let name: String
    let instanceKey: String
    let isActive: Bool
    let createdAt: Date
    let updatedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, userId, name, instanceKey, isActive
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }
}

@MainActor
actor InstanceService {
    static let shared = InstanceService()

    private let supabaseUrl: String
    private let backendProxyUrl: String
    private let auth = SupabaseAuthService.shared

    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    init(supabaseUrl: String = "", backendProxyUrl: String = "") {
        self.supabaseUrl = supabaseUrl.isEmpty
            ? Bundle.main.infoDictionary?["SUPABASE_URL"] as? String ?? "https://supabase.com"
            : supabaseUrl
        self.backendProxyUrl = backendProxyUrl.isEmpty
            ? Bundle.main.infoDictionary?["BACKEND_PROXY_URL"] as? String ?? "\(supabaseUrl)/functions/v1/mobile-instance-api"
            : backendProxyUrl
    }

    func createInstance(name: String, instanceKey: String) async throws -> Instance {
        guard case .signedIn(let session) = auth.authState else {
            throw NSError(domain: "InstanceService", code: 401, userInfo: [NSLocalizedDescriptionKey: "Not signed in"])
        }

        // CRITICAL FIX 5.1: Use backend proxy instead of direct Supabase API call
        guard let url = URL(string: backendProxyUrl) else {
            throw NSError(domain: "InstanceService", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }

        let body: [String: AnyCodable] = [
            "name": .string(name),
            "instance_key": .string(instanceKey)
        ]

        let bodyData = try JSONSerialization.data(withJSONObject: body.mapValues { value -> Any in
            switch value {
            case .string(let s):
                return s
            case .int(let i):
                return i
            case .bool(let b):
                return b
            default:
                return NSNull()
            }
        })

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(session.access_token)", forHTTPHeaderField: "Authorization")
        request.httpBody = bodyData

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "InstanceService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 201 else {
            throw NSError(domain: "InstanceService", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResponse.statusCode)"])
        }

        let instance = try decoder.decode(Instance.self, from: data)
        return instance
    }

    func fetchInstances() async throws -> [Instance] {
        guard case .signedIn(let session) = auth.authState else {
            throw NSError(domain: "InstanceService", code: 401, userInfo: [NSLocalizedDescriptionKey: "Not signed in"])
        }

        // CRITICAL FIX 5.1: Use backend proxy instead of direct Supabase API call
        let urlString = "\(backendProxyUrl)?user_id=\(session.user.id)"
        guard let url = URL(string: urlString) else {
            throw NSError(domain: "InstanceService", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(session.access_token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "InstanceService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 200 else {
            throw NSError(domain: "InstanceService", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResponse.statusCode)"])
        }

        let instances = try decoder.decode([Instance].self, from: data)
        return instances
    }

    func deleteInstance(id: String) async throws {
        guard case .signedIn(let session) = auth.authState else {
            throw NSError(domain: "InstanceService", code: 401, userInfo: [NSLocalizedDescriptionKey: "Not signed in"])
        }

        // CRITICAL FIX 5.1: Use backend proxy instead of direct Supabase API call
        let urlString = "\(backendProxyUrl)?id=\(id)"
        guard let url = URL(string: urlString) else {
            throw NSError(domain: "InstanceService", code: 400, userInfo: [NSLocalizedDescriptionKey: "Invalid URL"])
        }

        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(session.access_token)", forHTTPHeaderField: "Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw NSError(domain: "InstanceService", code: 0, userInfo: [NSLocalizedDescriptionKey: "Invalid response"])
        }

        guard httpResponse.statusCode == 200 else {
            throw NSError(domain: "InstanceService", code: httpResponse.statusCode, userInfo: [NSLocalizedDescriptionKey: "HTTP \(httpResponse.statusCode)"])
        }
    }
}

File: android/app/src/main/java/com/helix/services/InstanceService.kt
/**
 * Instance Service - Helix Android
 * Manages instance creation and fetching from Supabase
 */

package com.helix.services

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

@Serializable
data class Instance(
    val id: String,
    val user_id: String,
    val name: String,
    val instance_key: String,
    val is_active: Boolean,
    val created_at: String,
    val updated_at: String? = null
)

sealed class ApiResult<T> {
    data class Success<T>(val data: T) : ApiResult<T>()
    data class Error<T>(val message: String, val code: Int? = null) : ApiResult<T>()
}

class InstanceService(
    private val supabaseUrl: String,
    private val authTokenProvider: suspend () -> String?
) {
    // CRITICAL FIX 5.1: Use backend proxy URL instead of direct Supabase
    private val backendProxyUrl = "$supabaseUrl/functions/v1/mobile-instance-api"

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json { ignoreUnknownKeys = true }

    suspend fun createInstance(userId: String, name: String, instanceKey: String): ApiResult<Instance> =
        withContext(Dispatchers.IO) {
            try {
                val authToken = authTokenProvider() ?: ""
                // CRITICAL FIX 5.1: Use backend proxy instead of direct Supabase API call
                val requestBody = json.encodeToString(mapOf(
                    "name" to name,
                    "instance_key" to instanceKey
                )).toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url(backendProxyUrl)
                    .post(requestBody)
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer $authToken")
                    .build()

                val response = httpClient.newCall(request).execute()

                if (!response.isSuccessful) {
                    return@withContext ApiResult.Error<Instance>(
                        "HTTP ${response.code}",
                        response.code
                    )
                }

                val body = response.body?.string() ?: ""
                val responseBody = json.decodeFromString<Map<String, Instance>>(body)
                val instance = responseBody["data"] ?: throw Exception("No data in response")
                ApiResult.Success(instance)
            } catch (e: Exception) {
                ApiResult.Error(e.message ?: "Unknown error")
            }
        }

    suspend fun fetchInstances(userId: String): ApiResult<List<Instance>> =
        withContext(Dispatchers.IO) {
            try {
                val authToken = authTokenProvider() ?: ""
                // CRITICAL FIX 5.1: Use backend proxy instead of direct Supabase API call
                val url = "$backendProxyUrl?user_id=$userId"

                val request = Request.Builder()
                    .url(url)
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer $authToken")
                    .build()

                val response = httpClient.newCall(request).execute()

                if (!response.isSuccessful) {
                    return@withContext ApiResult.Error<List<Instance>>(
                        "HTTP ${response.code}",
                        response.code
                    )
                }

                val body = response.body?.string() ?: "[]"
                val responseBody = json.decodeFromString<Map<String, List<Instance>>>(body)
                val instances = responseBody["data"] ?: emptyList()
                ApiResult.Success(instances)
            } catch (e: Exception) {
                ApiResult.Error(e.message ?: "Unknown error")
            }
        }

    suspend fun deleteInstance(instanceId: String): ApiResult<Unit> =
        withContext(Dispatchers.IO) {
            try {
                val authToken = authTokenProvider() ?: ""
                // CRITICAL FIX 5.1: Use backend proxy instead of direct Supabase API call
                val url = "$backendProxyUrl?id=$instanceId"

                val request = Request.Builder()
                    .url(url)
                    .delete()
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer $authToken")
                    .build()

                val response = httpClient.newCall(request).execute()

                if (!response.isSuccessful) {
                    return@withContext ApiResult.Error<Unit>(
                        "HTTP ${response.code}",
                        response.code
                    )
                }

                ApiResult.Success(Unit)
            } catch (e: Exception) {
                ApiResult.Error(e.message ?: "Unknown error")
            }
        }
}
